File: Data Structures & Algorithms/pow-x-n/submission-0.py
# 2^1
# 2^1
class Solution:
    def myPowPos(self, x, n):
        if n == 0: return 1
        
        ans = self.myPow(x, int(n/2))
        res = 0

        if n % 2 == 0:
            res = ans * ans
        else:
            res = ans * ans * x

        return res

    def myPow(self, x: float, n: int) -> float:
        ans = self.myPowPos(x, abs(n))
        if n < 0: return 1/ans
        return ans

File: Data Structures & Algorithms/asteroid-collision/submission-0.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> st;

        for (int a:asteroids) {
            if (a > 0) {
                st.push(a);
                continue;
            }
            bool addA = true;
            while (st.size() != 0 && st.top() > 0 && a < 0) {
                int b = st.top();

                if (abs(a) == b) {
                    st.pop();
                    addA = false;
                    break;
                } else if (abs(a) < b) {
                    addA = false;
                    break;
                } else {
                    st.pop();
                    addA = true;
                }
            }
            if (addA) {
                st.push(a);
            }
        }
        vector<int> ans;
        while(!st.empty()) {
            int x = st.top();
            ans.push_back(x);
            st.pop();
        } 
        reverse(ans.begin(), ans.end());
        return ans;
    }
};